### PR TITLE
implement the functionality for constructing the PKI for virtual cluster

### PR DESCRIPTION
**Pull request by charleszheng44**
1. implement the functionality for constructing the PKI
2. create sample yaml for the ClusterVersion

### CODE REVIEW

File: incubator/virtualcluster/config/rbac/auth_proxy_service.yaml [hunk 1]
```diff
@@ -9,7 +9,7 @@ metadata:
     control-plane: controller-manager
     controller-tools.k8s.io: "1.0"
   name: controller-manager-metrics-service
-  namespace: system
+  namespace: kube-system
```
Comment (Fei-Guo) [inline incubator/virtualcluster/config/rbac/auth_proxy_service.yaml]: change to virtualcluster-system

File: incubator/virtualcluster/pkg/controller/add_clusterversion.go [hunk 2]
```diff
@@ -16,11 +16,11 @@ limitations under the License.
...
 
-import (
-	"github.com/multi-tenancy/incubator/virtualcluster/pkg/controller/clusterversion"
-)
-
-func init() {
-	// AddToManagerFuncs is a list of functions to create controllers and add them to a manager.
-	AddToManagerFuncs = append(AddToManagerFuncs, clusterversion.Add)
-}
+// import (
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/controller/add_clusterversion.go]: why do you comment out these?

File: incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go [hunk 3]
```diff
@@ -68,13 +111,266 @@ var _ reconcile.Reconciler = &ReconcileVirtualcluster{}
 // ReconcileVirtualcluster reconciles a Virtualcluster object
 type ReconcileVirtualcluster struct {
 	client.Client
-	scheme *runtime.Scheme
+	scheme          *runtime.Scheme
+	clusterVersions map[string]*tenancyv1alpha1.ClusterVersion
+	virtualClusters map[string]*tenancyv1alpha1.Virtualcluster
+}
+
+func (r *ReconcileVirtualcluster) checkCRKind(request *reconcile.Request) (runtime.Object, CRKind, error) {
```
Comment (Fei-Guo) [inline incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: You cannot prevent vc and clusterversion from having same name which breaks this method. I suggest to use another reconcile loop to reconcile for clusterversion.

Comment (charleszheng44) [reply incubator/virtualcluster/pkg/controller/virtualcluster/virtualcluster_controller.go]: when a new vc cr is created, we have to refer to existing cluster versions. This cannot be achieved by using two separate controllers. In other words, without a client set, we can’t query apiserver when we want to get all available clusterversions.

File: incubator/virtualcluster/pkg/controller/add_clusterversion.go (hunk 2 above)
Comment (charleszheng44) [reply incubator/virtualcluster/pkg/controller/add_clusterversion.go]: I have uncommented these lines

File: incubator/virtualcluster/config/rbac/auth_proxy_service.yaml (hunk 1 above)
Comment (charleszheng44) [reply incubator/virtualcluster/config/rbac/auth_proxy_service.yaml]: I think we should keep it this way. This service is for meta cluster.

Comment (Fei-Guo) [reply incubator/virtualcluster/config/rbac/auth_proxy_service.yaml]: ok

**Pull request merged**